### PR TITLE
change and optimize index struct

### PR DESCRIPTION
**Pull request by bbbearxyz**
1. split index struct to unique index/secondary index
2. unique index can work on the tae engine

## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #6842

## What this PR does / why we need it:
optimize index struct, now in the compute layer,  just one index struct.
now unique index only work on the tae engine, don't support on the 
memory engine, I will give a pr after this. (beacuse memory engine will 
report error)

### CODE REVIEW

**Review by daviszhen — APPROVED** [review]

**Review by reusee — APPROVED** [review]

**Review by fengttt — COMMENTED** [review]
no test at all?

**Comment by bbbearxyz** [comment]
> no test at all?

the pr is change and optimzie index struct, in the past, we have many tests for unique index, so I don't add any test.

**Comment by fengttt** [comment]
> > no test at all?
> 
> the pr is change and optimzie index struct, in the past, we have many tests for unique index, so I don't add any test.

really?   what is your code coverage?   resubmit this change when your code coverage reaches 80%

**Comment by fengttt** [comment]
> > > no test at all?
> > 
> > 
> > the pr is change and optimzie index struct, in the past, we have many tests for unique index, so I don't add any test.
> 
> really? what is your code coverage? resubmit this change when your code coverage reaches 80%

After some discussion, let say that can you run ut/bvt, look at the coverage using eyeballs, and update the commit message that most of the code is indeed covered by ut/bvt?

File: pkg/sql/plan/build_insert.go [hunk 1]
```diff
@@ -368,33 +368,19 @@ func getInsertTable(stmt tree.TableExpr, ctx CompilerContext) (*ObjectRef, *Tabl
...
-				for _, col := range tableDef.Cols {
-					info.Cols = append(info.Cols, col)
-					info.ColNames = append(info.ColNames, col.Name)
-				}
-				infos = append(infos, info)
-
+		if idxDef, ok := def.Def.(*plan.TableDef_DefType_UIdx); ok {
+			uIdxDef = idxDef.UIdx
+			for _, f := range uIdxDef.Fields {
+				println("bbb index def", f.Cols[0].Name)
```
Comment (qingxinhome) [inline pkg/sql/plan/build_insert.go:378]: The 'println' function can only appear in the service startup phase and debugging. After the service runs, it cannot be written like this。

**Review by qingxinhome — CHANGES_REQUESTED** [review]
The 'println' function can only appear in the service startup phase and debugging. After the service runs, it cannot be written like this

File: pkg/sql/plan/build_insert.go (hunk 1 above)
Comment (bbbearxyz) [reply pkg/sql/plan/build_insert.go:378]: my fault, I will delete this.

**Review by XuPeng-SH — APPROVED** [review]
lgtm

**Review by ouyuanning — APPROVED** [review]

File: pkg/vm/engine/types.go [hunk 2]
```diff
@@ -121,25 +121,35 @@ type ViewDef struct {
...
-type ComputeIndexDef struct {
-	IndexNames []string
-	TableNames []string
-	Uniques    []bool
-	Fields     [][]string
+type UniqueIndexDef struct {
+	IndexNames  []string
+	TableNames  []string
+	TableExists []bool
+	Parts       [][]string
```
Comment (nnsgmsone) [inline pkg/vm/engine/types.go:128]: What does parts mean?

File: pkg/sql/util/index_util.go [hunk 3]
```diff
@@ -55,7 +55,9 @@ func BuildUniqueKeyBatch(vecs []*vector.Vector, Attrs []string, p []*plan.ColDef
 		Attrs: make([]string, 1),
 		Vecs:  make([]*vector.Vector, 1),
 	}
-	if p[0].IsCPkey {
+	println("bbb build", len(p), p[0].Name)
```
Comment (daviszhen) [inline pkg/sql/util/index_util.go:58]: remove the println

**Review by daviszhen — CHANGES_REQUESTED** [review]

**Comment by bbbearxyz** [comment]
> > > > no test at all?
> > > 
> > > 
> > > the pr is change and optimzie index struct, in the past, we have many tests for unique index, so I don't add any test.
> > 
> > 
> > really? what is your code coverage? resubmit this change when your code coverage reaches 80%
> 
> After some discussion, let say that can you run ut/bvt, look at the coverage using eyeballs, and update the commit message that most of the code is indeed covered by ut/bvt?

ok, I already add ut for index_util.

**Review by daviszhen — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

File: pkg/frontend/session.go [hunk 4]
```diff
@@ -1677,20 +1676,67 @@ func (tcc *TxnCompilerContext) Resolve(dbName string, tableName string) (*plan2.
 					Partition: p,
 				},
 			})
-		} else if indexDef, ok := def.(*engine.ComputeIndexDef); ok {
-			fields := make([]*plan.Field, len(indexDef.Fields))
-			for i := range indexDef.Fields {
+		} else if indexDef, ok := def.(*engine.UniqueIndexDef); ok {
```
Comment (fengttt) [inline pkg/frontend/session.go:1679]: Why do we want to do runtime type cast?    can all these indexdef implement a common, well defined interface?

File: pkg/vm/engine/tae/catalog/schema.go [hunk 5]
```diff
@@ -1061,9 +1062,29 @@ func GetAttrIdx(attrs []string, name string) int {
 	panic("logic error")
 }
 
-type ComputeIndexInfo struct {
-	Name      string
-	TableName string
-	Unique    bool
-	Field     []string
+type UniqueIndexInfo struct {
```
Comment (aptend) [inline pkg/vm/engine/tae/catalog/schema.go:1065]: please consider storing bytes of these structs in schema, and deserializing them when fetching.

**Review by fengttt — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Review by qingxinhome — APPROVED** [review]

**Review by aptend — APPROVED** [review]

**Pull request merged**